draft-ietf-tls-esni.md

```diff
@@ -822,22 +822,27 @@ error code.
 
 ### Handling HelloRetryRequest {#client-hrr}
 
-When the server sends a HelloRetryRequest, the client determines if ECH was
-accepted by checking the message for an "encrypted_client_hello" extension with
-an 8-byte payload equal to `hrr_accept_confirmation` as defined in
-{{backend-server}}. If found, the client presumes acceptance and handles the
-HelloRetryRequest using ClientHelloInner. Otherwise, it presumes rejection and
-handles the HelloRetryRequest using ClientHelloOuter. Note that the
-client-facing server does not send "encrypted_client_hello" in case of
+When the server sends a HelloRetryRequest, the client checks for the presence of
+an "encrypted_client_hello" extension. If none is found, then the client
+presumes rejection and handles the HelloRetryRequest using ClientHelloOuter.
+(Note that the client-facing server does not send this extension in its
+HelloRetryRequest.) Otherwise it proceeds as follows.
+
+If the extension's payload has a length other than 8, then the client aborts the
+handshake with an "illegal_parameter" alert. If the payload length is equal to
+8, then the client checks if the payload is equal to `hrr_accept_confirmation`
+as defined in {{backend-server}}. If so, then it presumes acceptance and handles
+handles the HelloRetryRequest using ClientHelloInner. Otherwise, it presumes
 rejection.
 
+[[OPEN ISSUE: Depending on what we do for issue#450, it may be appropriate to
+change the client behavior if the HRR payload is missing or malformed.]]
+
 The client encodes the second ClientHelloInner as in {{encoding-inner}}, using
 the second ClientHelloOuter for any referenced extensions. It then encrypts
 the new EncodedClientHelloInner value as a second message with the previous
 HPKE context as described in {{encrypting-clienthello}}.
 
-[[OPEN ISSUE: See https://github.com/tlswg/draft-ietf-tls-esni/issues/450.]]
-
 ## GREASE ECH {#grease-ech}
```

Comment:
If we allow the server to send HRR.ech with invalid value as rejection, does that mean that, by sending GREASE ECH, the client is obligated to handle this case? Do we need some text here? (Like the client checks HRR.ECH syntactically but interprets all values as rejection.)

Reply:
Yeah, I think the client is obligated to handle this case. In particular, it should parse the extension. (Isn't the server obligated to do the same for the GREASE ECH extension?)

Reply:
Right. What I'm saying is the server specification doesn't distinguish GREASE ECH from normal ECH that fails to decrypt (and one hopes it cannot!). But the client specification has a big branch between GREASE ECH and normal ECH, so we probably need text here if we want that to work. (And we need a big branch because a GREASE ECH client cannot actually run the steps to compute the confirmation. There's no ClientHelloInner or anything.)

Reply:
Do you think there needs to be text around this in this PR? I would think it would make sense to address this once we have text for #450.

Reply:
It seems somewhat separable. We can just say that, if the client send GREASE ECH, then it should always assume rejection and use CHOuter.